src/renderer/components/settings-execution.tsx

```diff
@@ -12,29 +14,76 @@ import * as React from 'react';
 import { IPackageManager } from '../npm';
 import { AppState } from '../state';
 
+export enum SettingItemType {
+  EnvVars = 'environmentVariables',
+  Flags = 'executionFlags',
+}
+
 interface ExecutionSettingsProps {
   appState: AppState;
 }
 
+interface ExecutionSettingsState {
+  [setting: string]: Record<string, string>;
+}
+
 /**
  * Settings content to manage execution-related preferences.
  *
  * @class ExecutionSettings
  * @extends {React.Component<ExecutionSettingsProps, {}>}
  */
 @observer
-export class ExecutionSettings extends React.Component<ExecutionSettingsProps> {
+export class ExecutionSettings extends React.Component<
+  ExecutionSettingsProps,
+  ExecutionSettingsState
+> {
   constructor(props: ExecutionSettingsProps) {
     super(props);
 
+    this.state = {
+      executionFlags: Object.assign(
+        {},
+        ...props.appState.executionFlags.map((flag, idx) => {
+          return { [idx]: flag };
+        }),
+      ),
+      environmentVariables: Object.assign(
+        {},
+        ...props.appState.environmentVariables.map((envVar, idx) => {
+          return { [idx]: envVar };
+        }),
+      ),
+    };
+
     this.handleDeleteDataChange = this.handleDeleteDataChange.bind(this);
     this.handleElectronLoggingChange = this.handleElectronLoggingChange.bind(
       this,
     );
-    this.handleExecutionFlagChange = this.handleExecutionFlagChange.bind(this);
-    this.handleEnvironmentVariableChange = this.handleEnvironmentVariableChange.bind(
-      this,
-    );
+
+    this.handleSettingsItemChange = this.handleSettingsItemChange.bind(this);
+    this.addNewSettingsItem = this.addNewSettingsItem.bind(this);
+  }
+
+  public componentDidMount() {
+    const { environmentVariables, executionFlags } = this.state;
+
+    if (Object.keys(executionFlags).length === 0) {
+      this.addNewSettingsItem(SettingItemType.Flags);
+    }
+
+    if (Object.keys(environmentVariables).length === 0) {
+      this.addNewSettingsItem(SettingItemType.EnvVars);
+    }
+  }
+
+  public componentDidUpdate() {
+    const { appState } = this.props;
```

Comment:
This needs to be done here because a call to `setState` queues a _pending_ state transition and as such isn't synchronous. So - if we update `appState` with `this.state` directly after a call to `setState`, the state is nontrivially likely to be the previous state.